### PR TITLE
sys/shell: terminate shell on Ctrl-D

### PR DESCRIPTION
**Pull request by HendrikVE**
Ctrl-D was not caught in a special case so it was interpreted as a standard character. Handle it now the same way like EOF and terminate the shell instance.

Related to #10788

### CODE REVIEW

File: tests/shell/main.c [hunk 1]
```diff
@@ -82,6 +82,12 @@ int main(void)
     char line_buf[SHELL_DEFAULT_BUFSIZE];
 
     /* define own shell commands */
+    shell_run_once(shell_commands, line_buf, SHELL_DEFAULT_BUFSIZE);
+
+    puts("shell exited");
+
+    /* Restart the shell after the previous one exits, so that we can test
+     * Ctrl-D exit */
     shell_run(shell_commands, line_buf, SHELL_DEFAULT_BUFSIZE);
```
Comment (fjmolinas) [inline tests/shell/main.c:91]: I would run this in a loop instead.

Comment (fjmolinas) [reply tests/shell/main.c:91]: Also in native I get an infinite loop, can we do something simililar as for #13675?

**Comment by fjmolinas** [comment]
Sorry @HendrikVE I completely lost track of this one, can you rebase?

File: tests/shell/main.c (hunk 1 above)
Comment (HendrikVE) [reply tests/shell/main.c:112]: Two different behaviours are tested here. When we do ctrl-d in `shell_run_once` the execution returns to `main` of the test program. If we do ctrl-d in `shell_run` the shell will be respawn. I added a second test case for ctrl-d and comments to emphasize this behaviour. Do you still get an infinite loop?

**Comment by HendrikVE** [comment]
I rebased this PR :)

**Comment by fjmolinas** [comment]
Solves the issue, please squash @HendrikVE!

File: tests/shell/tests/01-run.py [hunk 2]
```diff
@@ -45,16 +45,17 @@
...
+    # test empty line input
     ('\n', PROMPT),
 
     # test simple word separation
     ('echo a string', '"echo""a""string"'),
     ('echo   multiple   spaces   between   argv', '"echo""multiple""spaces""between""argv"'),
     ('echo \t tabs\t\t processed \t\tlike\t \t\tspaces', '"echo""tabs""processed""like""spaces"'),
 
-    # test long line
+    # test unknown commands
```
Comment (aabadie) [inline tests/shell/tests/01-run.py:58]: This change is unrelated and the initial comment was correct (the purpose of the test line 58 is to test long lines). But the new comment could be kept and moved before line 62 (since it makes sense for this one).

**Comment by benpicco** [comment]
Only unrelated failures:

 - run_test/tests/mtd_flashpage/samr21-xpro:llvm (#14929)
 - run_test/tests/pkg_libhydrogen/samr21-xpro:llvm (#15066)

File: tests/shell/tests/01-run.py (hunk 2 above)
Comment (HendrikVE) [reply tests/shell/tests/01-run.py:58]: It's not really testing long lines here, because these test strings have a fixed length. The test done by calling `check_line_exceeded` is better suited for this purpose, because it makes use of the shell's buffer size automatically and writes characters beyond the buffer size. Maybe we should replace the long sequence of numbers by a shorter one or by something completely different instead?

Comment (aabadie) [reply tests/shell/tests/01-run.py:58]: > The test done by calling check_line_exceeded is better suited for this purpose, because it makes use of the shell's buffer size automatically and writes characters beyond the buffer size.

It this case is already covered by another test then I would suggest to remove this case and only keep 'unknown_command' one. @fjmolinas what do you think ?

Comment (fjmolinas) [reply tests/shell/tests/01-run.py:58]: Well its still an unrelated change, even if it makes sense, maybe you could split those into a separate commit @HendrikVE ?

Comment (HendrikVE) [reply tests/shell/tests/01-run.py:58]: It's in a separate commit now

**Comment by fjmolinas** [comment]
Looks good, I'll do a small test run and ACK.

**Comment by fjmolinas** [comment]
I ran the test on a collection of boards, only failure on `arduino-mega2560`, re ran and no issues https://ci.inria.fr/ci-riot-tribe/job/build-pipeline.jk/196/artifact/.

**Review by fjmolinas — APPROVED** [review]
ACK.
